src/basic/MAT_MAT_SHARED.hpp

```diff
@@ -0,0 +1,170 @@
+//~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~//
+// Copyright (c) 2017-20, Lawrence Livermore National Security, LLC
+// and RAJA Performance Suite project contributors.
+// See the RAJAPerf/COPYRIGHT file for details.
+//
+// SPDX-License-Identifier: (BSD-3-Clause)
+//~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~//
+
+///
+/// Matrix matrix multiplication with shared memory
+///
+///
+
+#ifndef RAJAPerf_Basic_MAT_MAT_SHARED_HPP
+#define RAJAPerf_Basic_MAT_MAT_SHARED_HPP
+
+#include "RAJA/RAJA.hpp"
+#include "common/KernelBase.hpp"
+
+#define TL_SZ 16
+
+#define MAT_MAT_SHARED_DATA_SETUP                                              \
+  Real_ptr A = m_A;                                                            \
+  Real_ptr B = m_B;                                                            \
+  Real_ptr C = m_C;
+
+//Clang has trouble with the RAJA_TEAM_SHARED macros believing
+//its always compiling for the device
+#define MAT_MAT_SHARED_BODY_0_CLANG_HIP_CPU                   \
```

Comment:
An additional detail, there seems to be confusion when compiling with clang + HIP. The compiler kept trying to invoke `__syncthreads()` in a CPU region even though we guard based on what we are compiling for.

Reply:
When doing the device compile pass hipcc/clang will put in the device versions of the macros everywhere, in device functions, host device functions, and host only functions. Then it will make sure that code is valid everywhere, that's fine for device and host device functions, but it is not ok for host only functions. Nvcc doesn't look at host only code when it does the device pass so it doesn't see these kind of problems.

Reply:
I'm going to pull your comment to the .hpp file :)

Reply:
Crap, that's a good point.  That means we're going to have similar (though worse) issues with nvc++ compiling cuda.  We may want to think about how to re-work our macro guards to avoid the differences.